enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,194 @@
+---
+title: MicroShift Greenboot
+authors:
+  - ggiguash
+reviewers:
+  - "@copejon"
+  - "@fzdarsky"
+  - "@majopela"
+  - "@dhellmann"
+  - "@oglok"
+  - "@zshi-redhat"
+  - "@pacevedom"
+approvers:
+  - "@dhellman"
+api-approvers:
+  - None
+creation-date: 2022-12-22
+last-updated: 2022-12-22
+tracking-link:
+  - https://issues.redhat.com/browse/USHIFT-319
+---
+
+# MicroShift Greenboot
+
+## Summary
+
+Assure successful software upgrades and system configuration changes on Edge
+Devices running MicroShift by using [greenboot](https://github.com/fedora-iot/greenboot),
+the Generic Health Check Framework for `systemd` on `rpm-ostree` based systems.
+
+## Motivation
+
+Serviceability of Edge Devices is often limited or non-existent, which makes it
+challenging to troubleshoot device problems following a failed software or
+operating system upgrade.
+
+This functionality benefits the users by reducing the risk of being locked out
+of an Edge Device when upgrades take place. Users will not experience
+significant interruption of service in case of a failed upgrade.
+
+### User Stories
+
+As a user, I want to operate my devices without interruptions so that I can
+have my work done.
+
+As a site manager, I want to ensure uninterrupted work of hundreds of edge
+devices so that I can fulfill the site deliverables.
+
+As a system administrator, I want to have a solution for failed software or
+operating system upgrades so that I can reduce an Edge Device downtime for the
+device users.
+
+As a developer, I want to have health check procedures for my software
+so that I can use them for ensuring that the software operates as designed.
+
+### Goals
+
+The goal of this enhancement is to describe techniques used for ensuring
+that an Edge Device running MicroShift functions properly following software
+upgrades or system configuration changes. If a failure is detected, the system
+is expected to boot into the last known working configuration using `rpm-ostree`
+rollback facilities.
+
+### Non-Goals
+
+The non-goals of this proposal are:
+- How to implement upgrade fallbacks on systems not based on `rpm-ostree`
+- How to implement upgrade fallbacks without using `greenboot`
+- How to upgrade Edge Devices, including best practices, etc.
+- How to monitor Edge Device software version and upgrade status
+- How to implement MicroShift data backup and restore for fallbacks
```

Comment:
isnt it a critical part of the solution?

Reply:
Yes, it's critical, but it will be handled in a separate enhancement proposal.

Reply:
I added references to the epics for follow-up.